Adjustable sequences

TobiasRasbold opened the pull request:
### Implements adjustable sequences for interval identifier

feasable sequence configurations:

- identical start and end marker
- using first start and first end marker
- using last start and last end marker
- using first start and last end marker (longest possible sequence:  #21 )
- using last start and first end marker (shortest possible sequence)

TobiasRasbold committed:
Merge pull request #1 from mansenfranzen/master

Update my fork from upstream repo

TobiasRasbold committed:
add interval notebook for development

TobiasRasbold committed:
add csv for testing

TobiasRasbold committed:
add transform function for first start, first end interval

TobiasRasbold committed:
add tests for transform_first_start_first_end interval identifier

TobiasRasbold committed:
begin interval identifier function for last start, last end

TobiasRasbold committed:
implement interval identifier function for last start, last end

TobiasRasbold committed:
add tests for interval_identifier last start, last end

TobiasRasbold committed:
add playground for new interval identifiers

TobiasRasbold committed:
formatting

TobiasRasbold committed:
adding spark first start last end interval identifier

TobiasRasbold committed:
adding spark first start last end interval identifier test data

TobiasRasbold committed:
notebook playground

TobiasRasbold committed:
add function _marker_bool_start_end

TobiasRasbold committed:
add marker_start_use and marker_end_use for using different intervals

TobiasRasbold committed:
add marker_start_use_first and marker_end_use_first to distinguish
sequence intervals

TobiasRasbold committed:
reorganize transform function

TobiasRasbold committed:
update tests using new transform implementation

TobiasRasbold committed:
update jupyter notebook

TobiasRasbold committed:
update jupyter notebook

mansenfranzen commented on src/pywrangler/pyspark/util.py:
```diff
@@ -52,3 +52,17 @@ def prepare_orderby(order_columns: TYPE_COLUMNS,
             for column, ascending in zipped]
+
+
+def prepare_reverse_orderby(order_columns: TYPE_COLUMNS,
```
DRY: reverse order could be an additional parameter of already existing `prepare_orderby` function because `prepare_reverse_orderby` and `prepare_orderby` are basically identical except for the reverse order.

TobiasRasbold committed:
streamline abstract variable names

TobiasRasbold committed:
update documentation

TobiasRasbold committed:
add tests for identical start/end markers
fixture for concrete interval identifier

TobiasRasbold committed:
update keyword arguments for interval identifier

mansenfranzen commented on src/pywrangler/pyspark/wranglers/interval_identifier.py:
```diff
@@ -105,3 +160,150 @@ def transform(self, df: DataFrame) -> DataFrame:
+        bool_valid = F.sum(bool_start_end).over(w_id) >= 2
+        valid_ids = F.when(bool_valid, ser_id).otherwise(0)
+
+        return df.withColumn(self.target_column_name, valid_ids).orderBy(orderby)
```
`.orderBy(orderby)` may trigger another shuffle which is not required because pyspark results do not need be ordered.

mansenfranzen commented on src/pywrangler/pyspark/wranglers/interval_identifier.py:
```diff
@@ -105,3 +160,150 @@ def transform(self, df: DataFrame) -> DataFrame:
+            .when(marker_col == self.marker_end, 0) \
+            .otherwise(None)
+
+        ffill_col = F.last(ffill, ignorenulls=True).over(w_lag.rowsBetween(-sys.maxsize, 0))
```
Spark [documentation](http://spark.apache.org/docs/latest/api/python/pyspark.sql.html?highlight=regex#pyspark.sql.Window.rowsBetween) suggests `Window.unboundedPreceding` instead of `-sys.maxsize` which may be more readable in the context of pyspark window functions.

mansenfranzen commented on src/pywrangler/pyspark/wranglers/interval_identifier.py:
```diff
@@ -105,3 +160,150 @@ def transform(self, df: DataFrame) -> DataFrame:
+            .otherwise(None)
+
+        ffill_col = F.last(ffill, ignorenulls=True).over(w_lag.rowsBetween(-sys.maxsize, 0))
+        # shift
```
Comments like `shift` / `add` / `negate` are kind of redundant given the supplied functions like `lag` etc. It may be more useful to comment on the intend of the statement (or multiple statements) even though it is admittedly not easy to describe what's going on here due to the complexity of the algorithm. 

As a suggestion, the logic of the specific algorithm could be provided within the doc string of the corresponding function.

mansenfranzen commented on src/pywrangler/pyspark/wranglers/interval_identifier.py:
```diff
@@ -105,3 +160,150 @@ def transform(self, df: DataFrame) -> DataFrame:
             .withColumn(self.target_column_name, renumerate_adjusted)
+
+    def _first_start_first_end(self, df: DataFrame) -> DataFrame:
+        """Extract interval ids from given dataframe.
```
Doc string needs to be adjusted. Preferably, it should contain a short description of the algorithm (as it was given for last start/first end). This applies to all new `transform` methods.

mansenfranzen commented on src/pywrangler/pyspark/wranglers/interval_identifier.py:
```diff
@@ -105,3 +160,150 @@ def transform(self, df: DataFrame) -> DataFrame:
+
+        # separate valid vs invalid: ids with start AND end marker are valid
+        bool_valid = F.sum(bool_start_end).over(w_id) >= 2
+        valid_ids = F.when(bool_valid, ser_id).otherwise(0)
```
**DRY**: Appears 3 times across different methods with slight variation of `>=` / `==`. Could be moved into separate helper function like `def _identify_valids(bool_start_end, w_id, ser_id, cmp)`

mansenfranzen commented on src/pywrangler/pyspark/wranglers/interval_identifier.py:
```diff
@@ -105,3 +160,150 @@ def transform(self, df: DataFrame) -> DataFrame:
+                result: pyspark.sql.Dataframe
+                    Same columns as original dataframe plus the new interval id column.
+
+                """
```
Indent is off.

mansenfranzen commented on src/pywrangler/pyspark/wranglers/interval_identifier.py:
```diff
@@ -105,3 +160,150 @@ def transform(self, df: DataFrame) -> DataFrame:
+
+        # define window specs
+        orderby_reverse = util.prepare_reverse_orderby(self.order_columns, self.ascending)
+        orderby = util.prepare_orderby(self.order_columns, self.ascending)
```
`orderby` not required: see comment for line 258.

mansenfranzen commented on src/pywrangler/pyspark/wranglers/interval_identifier.py:
```diff
@@ -105,3 +160,150 @@ def transform(self, df: DataFrame) -> DataFrame:
+        # add
+        add_col = end_marker_null_col + shift_col_negated
+        # cumsum
+        ser_id = F.sum(add_col).over(w_lag)
```
**DRY**: Similar to lines 188-202. Should be moved into separate function. The only difference is `w_lag` and the order of `marker_start`/`marker_end` which both can be passed as parameters like `def _GoodNameToBeFound(w_lag, start_first=True)`

TobiasRasbold replied:
implemented with _prepare_iids

TobiasRasbold committed:
streamline concretize abstract interval identifier

mansenfranzen commented on src/pywrangler/pyspark/wranglers/interval_identifier.py:
```diff
@@ -105,3 +160,150 @@ def transform(self, df: DataFrame) -> DataFrame:
+            .otherwise(ffill)
+
+        cum_sum = F.sum(start_and_fill_shift) \
+            .over(w_lag)
```
**DRY**: Similar to lines 188-202/236-250 except for addition of `shift_col_negated`. Should be moved into separate function like `def _GoodNameToBeFound(w_lag, start_first=True, add_shift_col=True)`. Additionally, different variable naming convention is used here which should aligned to the two former cases.

TobiasRasbold replied:
implemented with _prepare_iids

mansenfranzen commented on src/pywrangler/pyspark/wranglers/interval_identifier.py:
```diff
@@ -105,3 +160,150 @@ def transform(self, df: DataFrame) -> DataFrame:
+            .otherwise(cum_sum)
+
+        bfill = F.first(ser_id, ignorenulls=True) \
+            .over(w_id.orderBy(orderby).rowsBetween(0, sys.maxsize))
```
May substitute `sys.maxsize` with `Window.unboundedFollowing`.

mansenfranzen commented on src/pywrangler/pyspark/wranglers/interval_identifier.py:
```diff
@@ -105,3 +160,150 @@ def transform(self, df: DataFrame) -> DataFrame:
+            .over(w_id.orderBy(orderby).rowsBetween(0, sys.maxsize))
+        return df.withColumn(self.target_column_name, bfill) \
+            .withColumn(self.target_column_name, F.when(F.isnull(self.target_column_name), 0) \
+                        .otherwise(cum_sum))
```
Lines 300 until end could be slightly rewritten for better readability like: 

```python
        # delete noise in groups
        cols = [self.marker_start, self.marker_end]
        condition = ~marker_col.isin(cols) & (ffill == 0)
        ser_id = F.when(~condition, cum_sum)

        # backwards fill
        window = w_id.orderBy(orderby).rowsBetween(0, sys.maxsize)
        bfill = F.first(ser_id, ignorenulls=True).over(window)
        df = df.withColumn(self.target_column_name, bfill)

        # fill zeros and cum sum
        condition = F.isnull(self.target_column_name)
        fill = F.when(condition, 0).otherwise(cum_sum)
        return df.withColumn(self.target_column_name, fill)
```

mansenfranzen commented on tests/pyspark/wranglers/test_interval_identifier.py:
```diff
@@ -318,3 +347,102 @@ def test_identical_start_end_marker(wrangler, marker, test_case, spark):
+    kwargs = {"order_columns": "order",
+              "groupby_columns": "groupby"}
+
+    sequence = VectorizedCumSum(marker_column="marker",
```
`VectorizedCumSum` should be replaced with `wrangler`. Same holds for line 406 and 439.

mansenfranzen commented on tests/pyspark/wranglers/test_interval_identifier.py:
```diff
@@ -318,3 +347,102 @@ def test_identical_start_end_marker(wrangler, marker, test_case, spark):
+        noise=marker["noise"],
+        target_column_name="iids")
+
+    print(f"Input: \n {test_input}")
```
Print statements should be removed if not used for debugging.

TobiasRasbold replied:
remove print statements

mansenfranzen commented on tests/test_data/interval_identifier.py:
```diff
@@ -395,4 +395,199 @@ def identical_start_end_with_invalids_unsorted(start, noise,
+    return _return_dfs(data, target_column_name)
+
+
+def first_start_last_end_end_begins(start, end, noise, target_column_name):
```
Test cases for `first_start_last_end` do not cover noise elements in between intervals (like a sequence of end, noise, noise, start). This should be included to guarantee that noises in between end and start are assigned 0.

TobiasRasbold replied:
expand tests

mansenfranzen reviewed (changes requested):
**Thanks for the great PR!**

Some important changes are still required (which are not mentioned in the review below):

- [x] pandas' interval identifier needs to be modified, too. At minimum, it should raise a `NotImplementedError` if start and end usage patterns are not implemented, yet. At best, pandas gains full functionality, too.
- [x] The `IntervalIdentifier` base class in `pywrangler/wranglers.py` requires a documentation update because it will now support any variation of start-end-marker patterns.
- [x] Tests need to be adjusted and supplied for the changed `IntervalIdentifier` interface in `pywrangler/wranglers.py`.

A minor change:

- [x]  `notebooks/interval.ipynb` and `notebooks/oscillation_timeseries.csv` should be removed

TobiasRasbold committed:
add check_dtpyes to test

TobiasRasbold committed:
ignore notebooks

TobiasRasbold committed:
add additional parameter for prepare orderby function

TobiasRasbold committed:
add additional parameter for prepare orderby function

TobiasRasbold committed:
rename ser_id to raw_iids
add function identify_valids with adjustable comparator
use pyspark boundary start and end values instead (-) sys.maxsize

TobiasRasbold committed:
update documentation

TobiasRasbold committed:
update documentation
code refactoring

TobiasRasbold committed:
align tests because of continuous renumeration of iids

TobiasRasbold committed:
expand tests for first_start_last_end interval

TobiasRasbold committed:
update documentation

TobiasRasbold committed:
code refactoring

TobiasRasbold committed:
add functions for Naive Identifier

TobiasRasbold committed:
remove folder 'notebooks'

mansenfranzen commented:
**Thanks for the continuous effort on this PR!**

Unfortunately, there are some uncovered test scenarios which cause several algorithms to fail. 

## Minimal reproducible example

Consider the following test data:

```python
from pywrangler.pyspark.wranglers import interval_identifier
from pyspark.sql import SparkSession
from pyspark.sql import functions as F
from pyspark.sql import Window
import pandas as pd


start= "start"
end= "end"
noise = "noise"

data = [[1,  1, end],
        [2,  1, end],
        [3,  1, end],
        [4,  1, start],
        [5,  1, start],
        [6,  1, noise],
        [7,  1, end],
        [8,  1, end],
        [9,  1, start],
        [10, 1, start],
        [11, 1, start]]

kwargs = dict(marker_column="signal", 
              marker_start=start, 
              marker_end=end, 
              order_columns=["order"],
              groupby_columns=["groupby"])

dfp = pd.DataFrame(data, columns=["order", "groupby", "signal"])

spark = SparkSession.builder.getOrCreate()
df = spark.createDataFrame(dfp)
df.show()
```

```
+-----+-------+------+
|order|groupby|signal|
+-----+-------+------+
|    1|      1|   end|
|    2|      1|   end|
|    3|      1|   end|
|    4|      1| start|
|    5|      1| start|
|    6|      1| noise|
|    7|      1|   end|
|    8|      1|   end|
|    9|      1| start|
|   10|      1| start|
|   11|      1| start|
+-----+-------+------+
```

## Expected result
Regardless of the chosen algorithm, it should always yield only one valid interval (with varying length depending on the algorithm).

Let's step through the different implementations.

### First Start - Last End

```python
wrangler = VectorizedCumSum(marker_start_use_first=True,
                            marker_end_use_first=False, 
                            **kwargs)

wrangler.transform(df).show()
```
```
+-----+-------+------+----+
|order|groupby|signal|iids|
+-----+-------+------+----+
|    1|      1|   end|   0|
|    2|      1|   end|   0|
|    3|      1|   end|   0|
|    4|      1| start|   1|
|    5|      1| start|   1|
|    6|      1| noise|   1|
|    7|      1|   end|   1|
|    8|      1|   end|   1|
|    9|      1| start|   2|
|   10|      1| start|   2|
|   11|      1| start|   2|
+-----+-------+------+----+
```
Fails to reject the start markers at the end.

### First Start - First End

```python
wrangler = VectorizedCumSum(marker_start_use_first=True,
                            marker_end_use_first=True, 
                            **kwargs)

wrangler.transform(df).show()
```
```
+-----+-------+------+----+
|order|groupby|signal|iids|
+-----+-------+------+----+
|    1|      1|   end|   0|
|    2|      1|   end|   0|
|    3|      1|   end|   0|
|    4|      1| start|   1|
|    5|      1| start|   1|
|    6|      1| noise|   1|
|    7|      1|   end|   1|
|    8|      1|   end|   0|
|    9|      1| start|   2|
|   10|      1| start|   2|
|   11|      1| start|   2|
+-----+-------+------+----+

```
Fails to reject the start markers at the end.


### Last Start - Last End

```python
wrangler = VectorizedCumSum(marker_start_use_first=False,
                            marker_end_use_first=False, 
                            **kwargs)

wrangler.transform(df).show()
```
```
+-----+-------+------+----+
|order|groupby|signal|iids|
+-----+-------+------+----+
|   11|      1| start|   0|
|   10|      1| start|   0|
|    9|      1| start|   0|
|    8|      1|   end|   1|
|    7|      1|   end|   1|
|    6|      1| noise|   1|
|    5|      1| start|   1|
|    4|      1| start|   0|
|    3|      1|   end|   2|
|    2|      1|   end|   2|
|    1|      1|   end|   2|
+-----+-------+------+----+

```
Fails to reject the end markers at the start.

### Last Start - First End

```python
wrangler = VectorizedCumSum(marker_start_use_first=False,
                            marker_end_use_first=True, 
                            **kwargs)

wrangler.transform(df).show()
```
```
+-----+-------+------+----+
|order|groupby|signal|iids|
+-----+-------+------+----+
|    1|      1|   end|   0|
|    2|      1|   end|   0|
|    3|      1|   end|   0|
|    4|      1| start|   0|
|    5|      1| start|   1|
|    6|      1| noise|   1|
|    7|      1|   end|   1|
|    8|      1|   end|   0|
|    9|      1| start|   0|
|   10|      1| start|   0|
|   11|      1| start|   0|
+-----+-------+------+----+

```
Works as expected.

## Considerations

First Start/First End and Last Start/Last End might be fixed easily by modifying `_identify_valids`, however First Start/Last End is special on its own. In my opinion, it still might be the best choice to rely on the basic functionality of Last Start / First End and preprocess the `marker_column` with following modifications:

- **First Start / First End**: Remove duplicated start markers. Keep only the first.
- **Last Start / Last End**: Remove duplicated end markers. Keep only the last.
- **First Start / Last End**: Remove duplicated start markers. Keep only the first. Remove duplicated end markers. Keep only the last.

### Here is a hackish implementation:

**First, drop noise and forward fill:**

```
w1 = Window.partitionBy("groupby").orderBy("order")
condition = F.col("signal").isin([start,end])
denoised = F.when(condition, F.col("signal"))
ffill = F.last(denoised, ignorenulls=True).over(w1.rowsBetween(Window.unboundedPreceding, 0))
df.withColumn("cleaned", ffill).show()
```
```
+-----+-------+------+-------+
|order|groupby|signal|cleaned|
+-----+-------+------+-------+
|    1|      1|   end|    end|
|    2|      1|   end|    end|
|    3|      1|   end|    end|
|    4|      1| start|  start|
|    5|      1| start|  start|
|    6|      1| noise|  start|
|    7|      1|   end|    end|
|    8|      1|   end|    end|
|    9|      1| start|  start|
|   10|      1| start|  start|
|   11|      1| start|  start|
+-----+-------+------+-------+
```
**Get first starts only:**

```python
shift_start = F.lag(ffill).over(w1)

condition1 = ffill == start
shifted = F.when(condition1, shift_start)
erase = shifted == F.col("signal")

df.withColumn("only_first_start", F.when(erase, F.lit(None)).otherwise(F.col("signal"))).show()
```

```
+-----+-------+------+----------------+
|order|groupby|signal|only_first_start|
+-----+-------+------+----------------+
|    1|      1|   end|             end|
|    2|      1|   end|             end|
|    3|      1|   end|             end|
|    4|      1| start|           start|
|    5|      1| start|            null|
|    6|      1| noise|           noise|
|    7|      1|   end|             end|
|    8|      1|   end|             end|
|    9|      1| start|           start|
|   10|      1| start|            null|
|   11|      1| start|            null|
+-----+-------+------+----------------+
```

**Get last ends only:**
```python
shift_end = F.lag(ffill, count=-1).over(w1)

condition1 = ffill == end
shifted = F.when(condition1, shift_end)
erase = shifted == F.col("signal")

df.withColumn("only_last_end", F.when(erase, F.lit(None)).otherwise(F.col("signal"))).show()
```
```
+-----+-------+------+-------------+
|order|groupby|signal|only_last_end|
+-----+-------+------+-------------+
|    1|      1|   end|         null|
|    2|      1|   end|         null|
|    3|      1|   end|          end|
|    4|      1| start|        start|
|    5|      1| start|        start|
|    6|      1| noise|        noise|
|    7|      1|   end|         null|
|    8|      1|   end|          end|
|    9|      1| start|        start|
|   10|      1| start|        start|
|   11|      1| start|        start|
+-----+-------+------+-------------+


```

Both preprocessing steps should be stackable and could be applied before the *modified* marker column is passed to the already existing implementation of Last Start / First End.

This approach would greatly reduce the already present complexity in the various given implementations by basically relying on one main implementation with 2 possible preprocessing steps.

Commit:
Modify VectorizedCumSum to deal with various duplicates of start/end
marker more generically. Preserve adjusted implementation.

Commit:
Fix bug which caused valid end markers to be removed. Adjust tests.

TobiasRasbold committed:
Merge branch 'master' of https://github.com/mansenfranzen/pywrangler

Commit:
Add preprocessing for duplicated markers. Refactor code to comply with
added preprocessing.

Commit:
Add experimental TEST_VARS for referencing spark session.

Commit:
Add marker usage specific tests for pandas and pyspark.

Commit:
Add VectorizedCumSumAdjusted.

Commit:
Add initial ColumnCacher implementation.

Commit:
Refactor and simplify VectorizedCumSumAdjusted.

Commit:
Add VectorizedCumSumAdjusted for tests.

Commit:
Add tests for raw and valid iids result types.

Commit:
Add tests for raw and valid iids result types for pyspark and pandas
computation engine.

Commit:
Add `result_type` parameter and improve doc strings.

Commit:
Make NaiveIterator truly naive and drop vectorized preprocessing.

Commit:
Minor refactoring and doc improvements.

Commit:
Add interface tests for new `result_type` parameter.

Commit:
Add tests for ColumnCacher and improve docs.

Commit:
Streamline naming convention for `wranglers.py`.

Commit:
Streamline naming convention for `orderby_columns`.

Commit:
Add tests for missing groupby/orderby.

Commit:
Add `ensure_column`. Refactor `prepare_orderby`.

Commit:
Streamline naming convention for `pytest_parametrize_testcases`. Add
`pytest_parametrize_kwargs`. Improve doc. Add tests.

Commit:
Remove reference for spark session for testing purposes.

Commit:
Use `Optional` instead of `Union[None, ...]`.

Commit:
Fix typo and add type annotations.

Commit:
Move marker specific tests into one test collection. Refactor and add
docs.

Commit:
Fix multiple values for keyword argument.

Commit:
Remove obsolete imports.

Commit:
Add test for forced ascending parameter.

Commit:
Add tests for `groupby` and `sort_values`.

Commit:
Add `_compute_valid_renumerated_iids` to remove code duplications.

Commit:
Add `retain_none` parameter.

Commit:
Use `ensure_iterable` and drop None check.

Commit:
Add doc strings.

Commit:
Update changelog.

mansenfranzen commented:
PR ready for merging. Fixes #16, #21 and #24.
